### ISSUE TITLE
RF scaled_dot_product_attention

### ISSUE THREAD

**Issue by albertz** [rwth-i6/returnn#1555]
Add `scaled_dot_product_attention` as a function to RF, and use it in our attention code. (Does this also work with `RelPosSelfAttention`?)

In case of PyTorch, wrap `torch.nn.functional.scaled_dot_product_attention`. That should be much more compute and memory efficient compared to the direct implementation. It uses FlashAttention or potentially a number of other efficient kernels. (Although on older GPUs, probably not. See [my question on PyTorch discussion form](https://discuss.pytorch.org/t/scaled-dot-product-attention/201649/).)

**Comment by NeoLegends** [comment]
I'm thinking about the implications of this right now. Do we deal with ONNX export of RF models at all? How would using `F.scaled_dot_product_attention` affect this? Does it need an equivalent replacement path using "traditional ops" during export? In i6_models code the MHSA implementation w/o positional encoding uses pytorch's optimized implementation, but I still have to check if we do any model surgery (and replace it w/ a traditional impl) before export.

**Comment by albertz** [comment]
We can simply use a simple fallback implementation when this is done for ONNX. I think we already have that in some other places? E.g. in `TorchBackend.full`:
```python
if torch.onnx.is_in_onnx_export():
    # onnx::ConstantOfShape (via torch.full) must get shape as int64.
    # https://github.com/rwth-i6/returnn/issues/1333#issuecomment-1607236783
    shape = [dim.long() if isinstance(dim, torch.Tensor) else dim for dim in shape]
```
Also in `TorchBackend.conv`.

**Comment by dorian-K** [comment]
Look at this code snippet:
https://github.com/rwth-i6/returnn/blob/9678032af5228d7e837371cb641439f8932ec6e7/returnn/frontend/attention.py#L249-L251

Before this is executed, `axis` is the (shared) spatial dimension of both query and key/value.
But for the key and value matrices, this spatial dimension is then replaced with `hist_dim`, and its size tensor depends on the dimension `axis`, which is no longer present because it has been replaced by `hist_dim` in the key and value matrices. So those matrices are not well-defined without the `axis` dimension which is only present in the query matrix. So this is a problem in rf.CausalSelfAttention.

This is not a problem with the current `rf.dot_attention` because the masked dimension doesn't matter in the `query @ key` matmul, and after that the result tensor has both the `hist_dim` and `axis` dimensions, and everything is well defined again. But if we want to use pytorch for the whole attention operation, we would need to convert these dimensions to a mask matrix before the matmul happens. But now, in theory, both the query and key spatial dimensions could have masking, this makes the logic confusing.
Honestly I'm not sure if I am understanding this all correctly myself.

But what we actually want to do, is use the `is_causal` parameter of the torch `scaled_dot_product_attention`, which reduces the computational time by 2x because all those unnecessary computations that are masked out anyway do not happen anymore. So we need to change the CausalSelfAttention code anyways.

**Comment by albertz** [comment]
> Before this is executed, `axis` is the (shared) spatial dimension of both query and key/value.
> But for the key and value matrices, this spatial dimension is then replaced with `hist_dim`, and its size tensor depends on the dimension `axis`, which is no longer present because it has been replaced by `hist_dim` in the key and value matrices. So those matrices are not well-defined without the `axis` dimension which is only present in the query matrix. So this is a problem in rf.CausalSelfAttention.

I still don't understand. What do you mean by "those matrices are not well-defined without the `axis` dimension"? The dimension (`hist_dim`) is fully defined, or not? Its size is also already calculated, or not?

Is there some specific exception you get? (Or expect to get?)

**Comment by dorian-K** [comment]
> I still don't understand. What do you mean by "those matrices are not well-defined without the `axis` dimension"? The dimension (`hist_dim`) is fully defined, or not? Its size is also already calculated, or not?
> 
> Is there some specific exception you get? (Or expect to get?)

The key matrix will have dimensions `[batch, ..., hist_dim, embed_dim]`. The `hist_dim` dimension has a size tensor with dimensions `[axis]`. But `axis` is not a dimension of the key matrix. So if you want to know if a certain element is part of the key matrix, or masked out, this is not possible because the `hist_dim` cannot be interpreted without an index for the `axis` dimension, but we do not have an index for the `axis` dimension because the key matrix does not have this dimension

**Comment by albertz** [comment]
> The key matrix will have dimensions `[batch, ..., hist_dim, embed_dim]`. The `hist_dim` dimension has a size tensor with dimensions `[axis]`. But `axis` is not a dimension of the key matrix. So if you want to know if a certain element is part of the key matrix, or masked out, this is not possible because the `hist_dim` cannot be interpreted without an index for the `axis` dimension, but we do not have an index for the `axis` dimension because the key matrix does not have this dimension

Yes, `axis` is not a dim of the key matrix. So the masking is not even properly defined. But it becomes defined when you add it. It actually must be added to properly apply the masking - without having it, you anyway cannot apply masking in a meaningful way. But actually you would never really need to apply masking on the key matrix.

You multiply the keys with the queries (code from `dot_attention`, adapted to your dim names):
```python
energy = rf.matmul(query, keys, reduce=embed_dim)
# energy: [batch, ..., hist_dim, axis]
att_weights = rf.softmax(energy, axis=hist_dim)
```
There is where you get back the `axis` dim. It comes from the `matmul` with the queries. And the masking only need to be applied on the energies (`energy` here). You never need to apply it to `keys`.

I still don't understand how this is now relevant for [`torch.nn.functional.scaled_dot_product_attention`](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html). I guess you want to know how to set `attn_mask` and/or maybe also `is_causal`? It should be:
```python
attn_mask = rf.compare_bc(rf.range_over_dim(hist_dim, device=query.device), "<", hist_dim.get_size_tensor(device=query.device))
# attn_mask: [axis, hist_dim]
```
(Note, in `dot_attention`, the dim is actually called `axis`, what we call `hist_dim` here.)
For `is_causal`, you could write code that checks whether the attention mask is a square matrix and a lower triangular matrix. I'm not sure though that such a check is necessary, when you provide the `attn_mask`. I think it anyway only computes the attention where necessary based on the mask?
